### ISSUE TITLE
agbplay v2.0 Wishlist

### ISSUE THREAD

**Issue by ipatix** [ipatix/agbplay#68]
# Overview

Since there are a couple of major things that I, and I believe also other, people would like to be implemented in agbplay I want to open this issue and collect things that I do plan to implement when I'm in the mood and I get some time again to work on agbplay.

# Graphical User Interface (GUI)

agbplay's development was started in a time where I was getting a lot into the Linux world and applications. Thus agbplay has to some misfortune got the curses UI that we have today. While it does have the odd advantage on easily being usable remotely (which I have used before) I have to accept that this is really not a very useful trait. I know that there are some experiments with GUI support done [here](https://github.com/ahigerd/agbplay-gui/) but I'd like to have the GUI somewhat be part of the main project.

This leads me to the plan of splitting the current program into 3 modules:

- GUI
- Headless program
- libagbplay

If you have any suggestions what the GUI and the headless version should be able to do, please let me know and I'll add it to a list. Also any ideas regarding the library interface would be welcome.

- Do we really need a UI that resembles classic Sappy (which agbplay was loosely based on)?

# New features

## Settings editor.

Currently settings have to be manually changed in the JSON. If you want to quickly change things that becomes rather cumbersome and I'd like this to be possible with the UI.

## gsflib support

The current gsf support via the python script is somewhat hacky, as it merely creates a songlist. The original ROM is still required. Since from my understanding GSF strictly still is a ROM with everything but sound stripped, it should be possible to implement GSF support. This could vastly improve usability as people can just load one of the many tagged releases.

Perhaps this also depends on the game profile feature below since I don't want to implement GSF editing. Thus you'd have to choose between importing GSF tags or using an existing profile.

## Multiple game profiles

With ROM hacks people will often use the same game identification for ROMs with different playlists, different samples or different sound driver configurations. I want a quick way to select which profile the current ROM uses to change all the settings which are currently bound to the 4 digit game IDs. Perhaps even automatic detection of the profiles via magic bytes in a ROM which the modder can place somewhere.

# Binary Releases

This somewhat goes in hand with getting rid of the terminal curses UI. This is what currently prevents agbplay from working reliably across platforms due to differences in Unicode support. Ideally I want users to just be able to download the program and run it.

# You have ideas and suggestions?

So if you have made it until here: Do you have any suggestion that is outside of what I've mentioned here? Feel free to discuss and leave your comments.

**Comment by kittertron** [comment]
This is kind of niche, but a VGM logger for PSG. Since the GBA's PSG is basically just the DMG PSG, GBA psg data can be logged into VGM, even in MAME. The use-case of this is really specific, but it is something I'd like to see added!

**Comment by ipatix** [comment]
What do you mean with logger? A log file?

**Comment by kittertron** [comment]
> What do you mean with logger? A log file?

I meant something that would be able to log a vgm file, which is a log of soundchip instructions that can be played back as audio, but now that I think about it I don't quite know if AGBPlay uses emulation for the PSG so I don't know if it'd be possible. Nevermind to that, sorry.

**Comment by ipatix** [comment]
I don't see much purpose for that. The PSGs on the GBA are not programmable via instructions. They only offer register reads and writes. In addition this would only give you playback for PSG and not PCM. I could probably be done to a certain extend, but I personally don't see much use for this.

**Comment by scr-trees** [comment]
Not sure how much work it would be, but it would be amazing to get DKC3's sound engine to output the crystal clear sound of AGBPlay. I'm not even sure what the issue is with that game. Either way, it's awesome that you're still working on this project.

**Comment by ipatix** [comment]
Just wanted to give a quick update on how things are currently going:

The barebones GUI now exists, but it still lacks a few features:

- [x] file export
- [x] about screen
- [x] settings editor
- [x] profile editor
- [ ] switching profiles
- [x] show current sound mode
- [x] chord analyzer
- [x] saving profile
- [x] playlist editing
- [x] logging
- [x] gsf support
- [x] graceful error handling (e.g. a failed song table scan currently crashes the program)

This does not yet include the new planned features such as:

- [ ] sound overrides
- [ ] emulator integration
- [x] headless mode (i.e. exporting from CLI)
- [ ] zip'ped profiles

General bugs to fix:

- [x] correctly handle varying buffer sizes of audio API in respect to agbplay's buffer size at lowest latency without lockups.
- [x] fix unicode encoding of strings in profiles (we should probably use std::wstring internally for windows compatibility)

**Comment by H-A-M-G-E-R** [comment]
For emulator integration, let agbplay generate only the PCM samples and the emulator generate only the CGB channels.

**Comment by ipatix** [comment]
> For emulator integration, let agbplay generate only the PCM samples and the emulator generate only the CGB channels.

Although I'd generally question this idea, implementing this would be non-trivial either way. This would require agbplay to stay perfectly in sync with the emulator. The emulator integration is currently planned by means of hooking all the MP2K library functions and discarding the original call. If the original call would be kept, this would certainly cause desync at some point, possibly for unpredictable reasons like game lag.

The only "proper" way of implementing support would be to hook on a channel level, but this is not really intended in the design of agbplay, since agbplay does not have things like a channel limit. Also sub 60 Hz timings would no longer be possible (which have been tightly integrated in agbplay since ever basically).

**Comment by H-A-M-G-E-R** [comment]
What if we do the reverse: emulate the song/sound in the emulator to find inconsistancies between agbplay and the emulated audio.
I would also like agbplay to dump the songs, voice groups, and samples.

**Comment by ipatix** [comment]
> What if we do the reverse: emulate the song/sound in the emulator to find inconsistancies between agbplay and the emulated audio.

This does not work be design, since there **will** be  incosistencies, simply due to the quality improvements.

> I would also like agbplay to dump the songs, voice groups, and samples in ASM format so I can inject them to another rom using the sappy engine or to troubleshoot.

I'd never say never, but I certainly won't implement this any time soon, since there already are tools that can do this. The people who do decomps frequently do this. I also have a program which does this (check m4a2s), but it is rather old and possibly has flaws which the decomp tools do not have.

**Comment by H-A-M-G-E-R** [comment]
only fix the "bad" inconsistencies like CGB channels sounding too loud and some instruments having a higher pitch

**Comment by ipatix** [comment]
Please stop repeating yourself and post comments in their respective issue. This issue is about agbplay v2.0 features, not general bug reports. Also, since you apparently did not get the message yet: If you keep writing in an "everybody does as I say" tone, I will block your account from further interaction.

**Comment by ipatix** [comment]
Just to list the last things I want to get fixed before the full release:

- [ ] correct handling if an error occurs on the playback thread.
- [ ] apply profile changes to playback without reloading the game.
- [ ] create/copy/delete profile buttons in profile editor.

**Comment by ipatix** [comment]
<img width="625" height="387" alt="Image" src="https://github.com/user-attachments/assets/ad633093-d5e8-4734-bfaa-58b388b84d18" />

Change the resamplers to "nearest". Be warned of the crunchy sound :D

**Comment by ipatix** [comment]
Because I found that certain games sound better with different settings. Games with high samplerates will generally sound more crisp with the sinc resampler. That one would sound a bit dull on low samplerate games though. The Nearest and linear interpolator are there mostly because they were there the first.

**Comment by ruby-R53** [comment]
there's already [this list](https://vgmpf.com/Wiki/index.php?title=Game_Boy_Advance_sound_drivers) showing which games use which

**Comment by ruby-R53** [comment]
> If it doesn't use Nintendo's sound driver MP2K, then I will not implement it.

may i know why? are the other ones too undocumented or so different that it'd be way too much work?

**Comment by ipatix** [comment]
> may i know why? are the other ones too undocumented or so different that it'd be way too much work?

agbplay was never designed to support other engines. A large portion of the program would require a rewrite. Documentation of other engines is partially available, but I'm not sure if it as throughouly researched as mp2k.

**Comment by ruby-R53** [comment]
fair enough i guess, a bit of a shame 'cos i can't find any player for GSFs that supports dumping the individual sound channels like agbplay does with just the games' `.gba`'s

**Comment by ipatix** [comment]
This is a limitation of GSF. GSF simply runs in an emulator and is limited to what an emulator can do. Your foobar2k/winamp/audacious plugin emulator, which does sound, has no idea that it is actually being used for just sound playback.

The reason agbplay can look into this is, because it is aware of the sound format being used. This does require quite a bit of work, so I hope you can understand why I do not see this within the scope of agbplay.

**Comment by ruby-R53** [comment]
oh wow, it all makes sense now

**Comment by ruby-R53** [comment]
> i myself can no longer find any Qt 5 packages on my distro's repositories besides QtOpenGL and i fear this package might get removed soon too

yeah, turns out there'll be a complete purge of all Qt5 packages in my distro's repositories on June 30th, and that QtOpenGL is very included on that list

**Comment by ipatix** [comment]
That should not be a problem. agbplay compiles fine with Qt 6. Though, there are a couple of function deprecation warnings, which I am aware of.

**Comment by ruby-R53** [comment]
well when i compiled it for the first time it did complain about QtOpenGL being missing, specifically its Qt 5 version, which i then had to install separately
maybe it's changed now

**Comment by ipatix** [comment]
Maybe it just does that because it is still installed.
